alacritty_terminal/src/tty/windows/child.rs

```diff
@@ -0,0 +1,173 @@
+                INFINITE,
+                WT_EXECUTEINWAITTHREAD | WT_EXECUTEONLYONCE,
+            )
+        };
+
+        if 0 == success {
+            Err(Error::last_os_error())
+        } else {
+            Ok(HandleWaitSignal { wait_handle: AtomicPtr::from(wait_handle) })
+        }
+    }
+}
+
+impl Drop for HandleWaitSignal {
+    fn drop(&mut self) {
+        unsafe {
+            // Cancel WinAPI wait for child process exiting
+            UnregisterWait(self.wait_handle.load(Ordering::Relaxed));
+        }
+    }
+}
+
+pub struct ChildProcessWatcher {
+    _on_exit: HandleWaitSignal,
+    events_rx: Receiver<ChildEvent>,
+}
+
+impl ChildProcessWatcher {
+    pub fn new(subprocess_handle: HANDLE) -> Result<ChildProcessWatcher, Error> {
+        let (sender, receiver) = channel();
+        let on_exit = HandleWaitSignal::new(subprocess_handle, move || {
+            let _ = sender.send(ChildEvent::Exited);
+        })?;
+
+        Ok(ChildProcessWatcher { _on_exit: on_exit, events_rx: receiver })
+    }
+
+    pub fn events_rx(&self) -> &Receiver<ChildEvent> {
+        &self.events_rx
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::io::Error;
+    use std::ptr;
+    use std::time::Duration;
+
+    use std::sync::mpsc::channel;
+
+    use widestring::U16CString;
+
+    use winapi::shared::ntdef::LPWSTR;
+    use winapi::um::processthreadsapi::{
+        CreateProcessW, TerminateProcess, PROCESS_INFORMATION, STARTUPINFOW,
+    };
+
+    use super::*;
+
+    fn make_cmd_process() -> Result<HANDLE, Error> {
+        let mut pi: PROCESS_INFORMATION = Default::default();
+        let mut si: STARTUPINFOW = Default::default();
+        let cmdline = U16CString::from_str("cmd.exe").unwrap();
+
+        unsafe {
+            if 0 == CreateProcessW(
+                ptr::null(),
+                cmdline.as_ptr() as LPWSTR,
+                ptr::null_mut(),
+                ptr::null_mut(),
+                false as i32,
+                Default::default(),
+                ptr::null_mut(),
+                ptr::null_mut(),
+                &mut si as *mut STARTUPINFOW,
+                &mut pi as *mut PROCESS_INFORMATION,
+            ) {
+                Err(Error::last_os_error())
+            } else {
+                Ok(pi.hProcess)
+            }
+        }
+    }
+
+    #[test]
+    pub fn on_handle_wait_runs_callback_when_process_exits() {
+        // The test should only take few milliseconds to succeed but will
+        // only fail due to timeout. It can be safely assumed that if callback
+        // has not been called after `WAIT_TIMEOUT` duration, there is some
+        // issue with the code under test.
+        const WAIT_TIMEOUT: Duration = Duration::from_millis(200);
+
+        let (sender, receiver) = channel::<()>();
+        let subprocess_handle = make_cmd_process().unwrap();
+
+        // Setup callback to signal Condvar when process exits
+        let _hnd_wait = HandleWaitSignal::new(subprocess_handle, move || {
```

Comment:
`_hnd_wait` isn't really a great variable name. I'd assume this is likely to change with the rename of `HandleWaitSignal` anyways though, so I don't want to make any concrete recommendations yet.